### ISSUE TITLE
Make print() builtin synchronous

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3028]
`print(@map)` is designed to print all entries in a map. It currently does this by passing a MapID reference to userspace, where bpftrace iterates over the map entries to print them out.

The problem is that this makes `print(@map)` an asynchronous operation, which can produce unexpected results:
```
# bpftrace -e 'BEGIN { @x[0] = 1; print(@x); @x[0] = 2; exit() }'
Attaching 1 probe...
@x[0]: 2
```

Now that BPF has support for iterating over map elements in the kernel, we will be able to reimplement `print(@map)` as a synchronous operation by writing all the map keys+values to the ring buffer instead of a single MapID reference.

I think this could be implemented with minimal new codegen once #3003 is merged - by rewriting the AST of any `print(@map)` call into `for ($kv : @map) { print($kv) }`. Some work will be needed to ensure the entries are still formatted and sorted correctly.

The previous script should produce a more expected result in the future:
```
# bpftrace -e 'BEGIN { @x[0] = 1; print(@x); @x[0] = 2; exit() }'
Attaching 1 probe...
@x[0]: 1
```

This also applies to some other builtin functions which operate on maps asynchronously: `clear()`, `zero()`

**Comment by danobi** [comment]
While it's true changing from async to sync is backwards compatible, I'm a little worried about large maps overflowing the perg/ring buffer. In theory the user could be relying on the ability to print a huge map.

We could always add some way for the user to opt in, but I also dislike these kinds of knobs in general.

Seems like a tradeoff and I don't know the right answer atm.

**Comment by danobi** [comment]
Tangentially related: rust and golang have a edition/version mechanisms for these kinds of subtle changes. Basically you tell the compiler which edition/version you want to build against, and compiler will maintain semantics of that version. Newer code, when you `cargo new` or `go mod init` will target the latest version.

A mechanism like that could help us here. But it's probably overkill. Something to think about though.

**Comment by danobi** [comment]
(Sorry for stream of consciousness)

Maybe we could add an `aprint()` alias for the old behavior? And `print()` gets sync by default. I imagine this is beneficial for nearly everyone while also giving old users a way out.

**Comment by ajor** [comment]
> Maybe we could add an `aprint()` alias for the old behavior? And `print()` gets sync by default. I imagine this is beneficial for nearly everyone while also giving pathological users a way out.

In general, I'm more in favour of adding new syntax instead of new bulitin functions as I think it tends to be more flexible. What about introducing a way to take references to maps and allowing them to be printed, e.g.:
```
$map_ref = &@mymap;
print($map_ref);
```

It could also be useful in other scenarios, e.g. for double buffering:
```
$map_ref = $cond ? &@map1 : &@map2;
$map_ref[blah] = foo;
```

**Comment by janet-campbell** [comment]
I don't think changing the default is a good idea, and I'd point out that what this will do to synchronous latency in a hot path is considerable.  While doing prints from a probe isn't ideal to begin with, making them synchronous has the potential to make them much less useful as a rapid profiling tool as the print latency may be multiplied.

I use relatively large maps and even though I'm printing, clearing and double buffering them from a timer probe, I think this could be a step back.

**Comment by jordalgo** [comment]
I'm leaning toward @danobi's suggestion of adding an async print builtin. It's more explicit and removes the need for map copying.

**Comment by janet-campbell** [comment]
To mention history, the dtrace printing functions (printf, printa) were buffered and asynchronous by default since its introduction 20 years ago.  While you can #pragma otherbuffertypes and get some variant behavior, the default has stayed the same for good reason: choosing something else may have severe consequences so you want it to be a conscious decision.

I'd argue that async printing by default is the principle of least surprise here.

**Comment by ajor** [comment]
@janet-campbell Maybe we're talking about different types of "asynchronous printing" here.

`printf` in bpftrace is an asynchronous function, in that calling it just involves putting data onto a ring buffer. Userspace asynchronously reads this ring buffer and prints out the data. This will never change.

The data which is displayed by `printf` is what I'm calling "synchronous" here though - actual values are put onto the ring buffer at the time `printf` is called, so what's displayed to the user can not change after this point. I'm not familiar enough with DTrace to say for certain, but I can't see it being useful to have this work any other way.

What would you expect this script to print? 
```
BEGIN
{
  @val = 1;
  printf("%d\n", @val);
  @val = 2;
}
```
Displaying "1" to the user is what I mean by synchronous printing, even though a ring buffer is still used.

The bpftrace `print()` function doesn't work like this, and that's what I'd like to improve.

I do see that there's a use for maintaining an asynchronous printing functionality, but it'd only be needed by advanced users using double buffering techniques who really care about latency. Advanced users can opt-in to this kind of thing.

The default should be usable and understandable by the average newcomer to bpftrace, whose main concern is whether their data appears to be "incorrect".

**Comment by danobi** [comment]
> > I'm leaning toward @danobi's suggestion of adding an async print builtin. It's more explicit and removes the need for map copying.
> 
> Why not the explicit map references approach ([#3028 (comment)](https://github.com/bpftrace/bpftrace/issues/3028#issuecomment-2010250450))? I really don't want to add new builtin functions to the language unnecessarily and explicit references appears more flexible to me.

Do you mind sharing your thought process behind preferring syntax extensions over functions? IMO functions are strongly typed and have clear bounds on interaction w/ other parts of the language. Whereas syntax extensions combinatorially increase complexity of language and may have unforeseen side effects.

Granted, in this case I could see a good argument for adding an address-of operator. We do need a way to allow passing larger objects (like maps and structs) to user defined functions as well. Assuming you are also thinking of pointer-of as the exact semantics.

I haven't thought through the implications of reference semantics yet, but presumably that would require a new type in the type system, which I think we are in agreement is best avoided if possible.

**Comment by janet-campbell** [comment]
@ajor I can see the whole least surprise thing.  I do think it would be good to comparison test it at scale to see what the ramifications are.

**Comment by jordalgo** [comment]
> Why not the explicit map references approach (https://github.com/bpftrace/bpftrace/issues/3028#issuecomment-2010250450)?

While maybe the underlying implementation might be the same between `aprint(@map)` and `print($map_ref)`, the syntax for printing map references doesn't really help the user understand what's happening. Why is printing of map ref sync and printing of the map itself async? References aren't snapshots.

While I agree that references are more flexible, I also agree with @danobi that introducing them into the language is non-trivial and would require a lot more testing and investigation - especially vs just adding `aprint` or something.

**Comment by ajor** [comment]
> Do you mind sharing your thought process behind preferring syntax extensions over functions? IMO functions are strongly typed and have clear bounds on interaction w/ other parts of the language. Whereas syntax extensions combinatorially increase complexity of language and may have unforeseen side effects.

Yes, functions are easier for us to implement, but I feel that they make for a less expressive programming environment for users. Adding a new function is only useful for this one specific use-case. Adding new syntax gives more power to user and lets them do things we might not even have thought of.

I believe the language should provide a minimal, useful set of primitives, which users can compose to build whatever functionality they want. We still have to decide what those primitives are, but I think in the past we've included too many non-composable functions as part of the core language (*cough* `join`).

> Granted, in this case I could see a good argument for adding an address-of operator. We do need a way to allow passing larger objects (like maps and structs) to user defined functions as well. Assuming you are also thinking of pointer-of as the exact semantics.

Things like this is precisely why I prefer syntax over functions! Syntax tends to solve many problems at once.



> While maybe the underlying implementation might be the same between `aprint(@map)` and `print($map_ref)`, the syntax for printing map references doesn't really help the user understand what's happening. Why is printing of map ref sync and printing of the map itself async? References aren't snapshots.

The opposite! I was suggesting that `print(@)` be synchronous and `print(&@)` be asynchronous. The reasoning being that the `print()` function can be thought of as putting some data onto the ring buffer. `print(@)` passes the map by value, so copies the current data onto the ring buffer. `print(&@)` puts a reference to the map on the ring buffer so is asynchronous.


> While I agree that references are more flexible, I also agree with @danobi that introducing them into the language is non-trivial and would require a lot more testing and investigation - especially vs just adding `aprint` or something.

When it comes to language syntax, whatever decision we make will be staying with us for a long time. I think we need to implement whatever the best solution is with a view to the long-term and not based on how much work it'll be to implement now.

**Comment by jordalgo** [comment]
> The opposite! I was suggesting that print(@) be synchronous and print(&@) be asynchronous. 

That's better but I still think it suffers from the same lack of intuitiveness.

>  I think we need to implement whatever the best solution is with a view to the long-term and not based on how much work it'll be to implement now.

I agree that long term solutions are obviously preferred but I'm not thinking so much of the amount of work but rather how quickly can we meet the needs of our users. Maybe references don't automatically fall into this category but there have definitely been design discussions that just get stuck in purgatory (👀 script args). Do we also think that `aprint` would only be used for maps? Maybe it has other uses e.g. symbolized stacks?

**Comment by ajor** [comment]
Just reposting from the above linked issue to consolidate discussion in this thread:

Named function arguments as an alternative way of handling synchronous vs asynchronous map printing:
```
print(@mymap); // defaults to synchronous map printing
print(@mymap, async=true); // option to print asynchronously
```

I think this is my favourite solution so far.

**Comment by danobi** [comment]
For the existing builtin functions that support optional parameters, should we also give them optional named parameters?
